src/examples/libpmemobj/hashmap/hashmap_rp.c

```diff
@@ -0,0 +1,700 @@
+
+	pmemobj_persist(pop, D_RW(hashmap_rebuild.entries), sz_alloc);
+
+	pmemobj_defer_free(pop, D_RW(hashmap)->entries.oid, &actv[actv_cnt++]);
+
+	pmemobj_set_value(pop, &actv[actv_cnt++],
+		&D_RW(hashmap)->entries.oid.pool_uuid_lo,
+		hashmap_rebuild.entries.oid.pool_uuid_lo);
+	pmemobj_set_value(pop, &actv[actv_cnt++],
+		&D_RW(hashmap)->entries.oid.off,
+		hashmap_rebuild.entries.oid.off);
+
+	assert(sizeof(actv) / sizeof(actv[0]) >= actv_cnt);
+	pmemobj_publish(pop, actv, actv_cnt);
+
+	return 0;
+
+rebuild_err:
+	pmemobj_cancel(pop, actv, actv_cnt);
+
+	if (hashmap_rp_debug)
+		free(swaps_array);
+
+	return -1;
+}
+
+/*
+ * hm_rp_create --  initializes hashmap state, called after pmemobj_create
+ */
+int
+hm_rp_create(PMEMobjpool *pop, TOID(struct hashmap_rp) *map, void *arg)
+{
+	struct hashmap_args *args = (struct hashmap_args *)arg;
+	uint32_t seed = args ? args->seed : 0;
+
+	hashmap_create(pop, map, seed);
+
+	return 0;
+}
+
+/*
+ * hm_rp_check -- checks if specified persistent object is an instance of
+ * hashmap
+ */
+int
+hm_rp_check(PMEMobjpool *pop, TOID(struct hashmap_rp) hashmap)
+{
+	return TOID_IS_NULL(hashmap) || !TOID_VALID(hashmap);
+}
+
+/*
+ * hm_tx_init -- recovers hashmap state, called after pmemobj_open.
+ * Since hashmap_rp is performing rebuild/insertion completely or not at all,
+ * function is dummy and simply returns 0.
+ */
+int
+hm_rp_init(PMEMobjpool *pop, TOID(struct hashmap_rp) hashmap)
+{
+	return 0;
+}
+
+/*
+ * hm_rp_insert -- rebuilds hashmap if necessary and wraps insert_helper.
+ * returns:
+ * - 0 if successful,
+ * - 1 if value already existed
+ * - -1 if something bad happened
+ */
+int
+hm_rp_insert(PMEMobjpool *pop, TOID(struct hashmap_rp) hashmap,
+		uint64_t key, PMEMoid value)
+{
+	if (D_RO(hashmap)->count + 1 >= D_RO(hashmap)->resize_threshold) {
+		uint64_t capacity_new = D_RO(hashmap)->capacity * 2;
+		if (hm_rp_rebuild(pop, hashmap, capacity_new) != 0)
+			return -1;
+	}
+
+	return insert_helper(pop, D_RW(hashmap), key, value);
+}
+
+/*
+ * hm_rp_remove -- removes specified key from the hashmap,
+ * returns:
+ * - key's value if successful,
+ * - OID_NULL if value didn't exist or if something bad happened
+ */
+PMEMoid
+hm_rp_remove(PMEMobjpool *pop, TOID(struct hashmap_rp) hashmap,
+		uint64_t key)
+{
+	const uint64_t pos = index_lookup(D_RO(hashmap), key);
+
+	if (pos == 0)
+		return OID_NULL;
+
+	struct entry *entry_p =
+		(struct entry *)pmemobj_direct(D_RW(hashmap)->entries.oid);
+	entry_p += pos;
+	PMEMoid ret = entry_p->value;
+
+	size_t actvcnt = 0;
+
+	struct pobj_action actv[5];
+
+	pmemobj_set_value(pop, &actv[actvcnt++], &entry_p->hash,
+		entry_p->hash | (1ll << 63));
+	pmemobj_set_value(pop, &actv[actvcnt++],
+		&entry_p->value.pool_uuid_lo, 0);
+	pmemobj_set_value(pop, &actv[actvcnt++], &entry_p->value.off, 0);
+	pmemobj_set_value(pop, &actv[actvcnt++], &entry_p->key, 0);
+	pmemobj_set_value(pop, &actv[actvcnt++], &D_RW(hashmap)->count,
+		--D_RW(hashmap)->count);
```

Comment:
you are modifying the D_RW(hashmap)->count through a redo log *AND* normally with `--`.
Did you mean to write:
`pmemobj_set_value(..., &D_RW(hashmap)->count, D_RO(hashmap)->count - 1);`
?